evalml/pipelines/pipeline_base.py

```diff
@@ -280,19 +266,31 @@ def _score_all_objectives(self, X, y, y_pred, y_pred_proba, objectives):
     def model_family(cls):
         "Returns model family of this pipeline template"""
         component_graph = copy.copy(cls.component_graph)
-        return handle_component_class(component_graph[-1]).model_family
+        if isinstance(cls.component_graph, list):
+            return handle_component_class(component_graph[-1]).model_family
+        else:
+            # TODO: Does this make sense
+            return ModelFamily.ENSEMBLE
```

Comment:
I'm not sure what the best way to determine a model family for non-linear pipelines would be. The main problems:

- A non-linear pipeline can have multiple estimators, should the model family reflect all of them or just the final one?
- Since this is a class property, we only have access to the dictionary describing the pipeline, which may not be in any sort of computation order. Without the instantiated ComponentGraph object, we have no clean way of knowing which estimator is the final one

Since this is ambiguous, I've opted to bundle these all under the ensemble model family but I'm looking for input on this!

Reply:
Interesting question! These are some of the places where we call `problem_type` at the pipeline level:

1. To [skip cv](https://github.com/alteryx/evalml/blob/main/evalml/automl/automl_search.py#L643) for ensemble pipelines in `_compute_cv_scores` in AutoMLSearch.
2. To [set allowed_model_families](https://github.com/alteryx/evalml/blob/main/evalml/automl/automl_search.py#L482) in AutoMLSearch after pipelines are created
3. To [describe](https://github.com/alteryx/evalml/blob/main/evalml/pipelines/pipeline_base.py#L157) a pipeline.
4. To [check](https://github.com/alteryx/evalml/blob/main/evalml/tests/automl_tests/test_automl_search_classification.py#L538) which pipelines are created by automl in tests
5. To [check](https://github.com/alteryx/evalml/blob/main/evalml/model_understanding/graphs.py#L652) the input pipeline has a decision tree as a final estimator in our model understanding utils
6. To [check](https://github.com/alteryx/evalml/blob/main/evalml/model_understanding/graphs.py#L450) we don't pass in baseline pipelines for partial dependence
7. To [determine](https://github.com/alteryx/evalml/blob/main/evalml/automl/automl_algorithm/iterative_algorithm.py#L115) which pipelines to add to our ensemble in IterativeAlgorithm
8. To [determine](https://github.com/alteryx/evalml/blob/main/evalml/automl/automl_algorithm/iterative_algorithm.py#L101) how the pipeline scores should be saved in IterativeAlgorithm

In all of these use cases, what's intended is to get the final estimator/component model family. It's interesting we access `model_family` for the instance, not the class.  I think we can avoid the computation order problem you bring up by making this a `property` instead. 

I think we should fix this before merge. I don't think returning `ModelFamily.Ensemble` is a good idea because in our codebase, that typically means a `StackedEnsembleClassifier` or `StackedEnsembleRegressor` is in the pipeline, which isn't guaranteed to be the case here.